### PR TITLE
ci: build otel with m32

### PR DESCRIPTION
**Pull request by dbolduc**
Motivated by #11156 

opentelemetry-cpp does not have a binary package we can just install. So we have to build the code using our m32 toolchain.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11196)

### CODE REVIEW

File: ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile [hunk 1]
```diff
@@ -100,6 +100,28 @@ RUN curl -sSL https://github.com/nlohmann/json/archive/v3.11.2.tar.gz | \
...
+        -DWITH_ABSEIL=ON \
+        -DBUILD_TESTING=OFF \
+        -DOPENTELEMETRY_INSTALL=ON \
+        -DCMAKE_CXX_COMPILER=g++ \
+        -DCMAKE_CXX_FLAGS=-m32 \
+        -DCMAKE_FIND_ROOT_PATH=/usr/ \
+        -DCMAKE_FIND_ROOT_PATH_MODE_PROGRAM=NEVER \
+        -DCMAKE_FIND_ROOT_PATH_MODE_LIBRARY=ONLY \
+        -DCMAKE_FIND_ROOT_PATH_MODE_INCLUDE=ONLY \
+        -H. -Bcmake-out && \
```
Comment (coryan) [inline ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile:121]: I think we started using `-S . -B cmake-out` a while ago?

File: ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile [hunk 2]
```diff
@@ -100,6 +100,28 @@ RUN curl -sSL https://github.com/nlohmann/json/archive/v3.11.2.tar.gz | \
     cmake --build cmake-out --target install -- -j ${NCPU:-4} && \
     ldconfig
 
+WORKDIR /var/tmp/build/opentelemetry
+RUN curl -sSL https://github.com/open-telemetry/opentelemetry-cpp/archive/v1.8.3.tar.gz | \
+    tar -xzf - --strip-components=1 && \
+    cmake \
+        -DCMAKE_CXX_STANDARD=20 \
```
Comment (coryan) [inline ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile:107]: Why is this compiled with C++20 but the other things in this file are not?

Comment (dbolduc) [reply ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile:107]: I need to match the standard that absl is compiled with. I do not know what it is, but I know that building opentelemetry-cpp with C++14 gave me this error:

```
/usr/bin/ld: /usr/local/lib/libopentelemetry_exporter_in_memory.so: undefined reference to `absl::lts_20220623::variant_internal::ThrowBadVariantAccess()'
```

I tried C++20, and it worked, so I moved on and didn't look back.

Comment (coryan) [reply ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile:107]: Ah.... Abseil is built with the default C++ version on that platform (C++17).  OpenTelemetry is probably forcing a -DCMAKE_CXX_STANDARD=... on us with no need to do so.

**Review by coryan — APPROVED** [review]

File: ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile (hunk 1 above)
Comment (dbolduc) [reply ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile]: Fixed.

File: ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile (hunk 2 above)
Comment (dbolduc) [reply ci/cloudbuild/dockerfiles/fedora-m32.Dockerfile]: Thanks, I switched the standard to C++17.

I also added some comments so we don't have to dig up this PR to know why certain choices were made.

**Pull request merged**